File: studies/cs50x/week2/pset2/readability/readability.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>
#include <math.h>

// Your program must prompt the user for a string of text (using get_string).
int main(void)
{
    //Get our data
    string s = get_string("Text: ");
    //Introduce our 3 variables - arrays, where we'll store our data as we parse through.
    float sentences = 0;
    float words = 0;
    float letters = 0;

    //Parse through every letter of our text
    for (int i = 0, n = strlen(s); i < n; i++)
    {
        //If an uppercase or lowercase letter, count it.
        if ((s[i] >= 65 && s[i] <= 90) || (s[i] >= 97 && s[i] <= 122))
        {
            letters += 1;
        }
        //If a '_.?!' count a new word.
        if (s[i] == 32 || s[i] == 33 || s[i] == 46 || s[i] == 63)
        {
            words += 1;
            //If one of the latter three, count a new sentence. Skip the following space.
            if (s[i] > 32)
            {
                sentences += 1;
                i += 1;
            }
        }
    }

    float L = letters * 100 / words;
    float S = sentences * 100 / words;
    float index = round(0.0588 * L - 0.296 * S - 15.8);

    if (index >= 16)
    {
        printf("Grade 16+\n");
    }
    else if (index < 1)
    {
        printf("Before Grade 1\n");
    }
    else
    {
        printf("Grade %i\n", (int) index);
    }
}

// Your program should count the number of letters, words, and sentences in the text.
// You may assume that a letter is any lowercase character from a to z or any uppercase character from A to Z,
// any sequence of characters separated by spaces should count as a word, and that any occurrence of a period,
// exclamation point, or question mark indicates the end of a sentence.

// Your program should print as output "Grade X" where X is the grade level computed by the Coleman-Liau formula, rounded to the nearest integer.

// If the resulting index number is 16 or higher (equivalent to or greater than a senior undergraduate reading level), your program should output "Grade 16+" instead of giving the exact index number. If the index number is less than 1, your program should output "Before Grade 1".

File: self_teaching/cs50x/week6/pset6/readability/readability.py
from cs50 import get_string
import re
import string

# List of U and L letters, then list of punctuation to match to later.
alphabet = string.ascii_uppercase[:26] + string.ascii_lowercase[:26]

punctuation = [".", "!", "?"]

# Get input of text from user.
text = get_string("Input text here: ")

# Introduce our variables, and tally as we iterate through.
letters = 0
# Account for the last word of the text not having a space after it.
words = 1
sentences = 0

for i in range(len(text)):
    if text[i] in alphabet:
        letters += 1
    if (text[i] == " "):
        words += 1
    if text[i] in punctuation:
        sentences += 1

l = letters * 100 / words
s = sentences * 100 / words
index = round(0.0588 * l - 0.296 * s - 15.8)

if index < 1:
    print("Before Grade 1")
elif index > 15:
    print("Grade 16+")
else:
    print(f"Grade {index}")

    # Recall that the Coleman-Liau index:
    # is computed as 0.0588 * L - 0.296 * S - 15.8,
    # L is the average number of letters per 100 words in the text
    # S is the average number of sentences per 100 words in the text.

    # Count the number of letters, words, and sentences in the text.
    # A letter is any lowercase character from a to z or any uppercase character from A to Z,
    # A word is any sequence of characters separated by spaces.
    # Any occurrence of a period, exclamation point, or question mark indicates the end of a sentence.

    # Your program should print as output "Grade X" where X is
    # the grade level computed by the Coleman-Liau formula, rounded to the nearest integer.
    # If the resulting index number is 16 or higher your program should output "Grade 16+"
    # If the index number is less than 1, your program should output "Before Grade 1".